lib/puppet/util/uri.rb

```diff
@@ -0,0 +1,14 @@
+# This module holds util funtions for URI's
+module Puppet::Util::Uri
+  # Mask credentials in given URI or address as string. Resulting string will
+  # contain '***' in place of password. It will only be replaced if actual
+  # password is given.
+  #
+  # @param uri [URI|String] an uri or address to be masked
+  # @return [String] a masked url
+  def mask_credentials(uri)
+    uri = URI.parse(uri) unless uri.is_a? URI
```

Comment:
The logic here should be a little different:

- if it's a URI, it should copy itself (then we should be altering the copy on the next line, not the URI directly)
- if it's a string, then create the URI from it (and still copy/modify as above)

Reply:
Yup. That's right. I noticed this problem after sending the change. I was supposed to fix it but it fell out of my head. I'll fix it.